In [51]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier,GradientBoostingClassifier,VotingClassifier
from sklearn.metrics import accuracy_score,classification_report,recall_score
from sklearn.model_selection import train_test_split

!pip install xgboost


[notice] A new release of pip is available: 25.3 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [52]:
import xgboost as xgb

In [53]:
df=pd.read_csv("novagen_dataset.csv")

In [54]:
df.head()

,Age,BMI,Blood_Pressure,Cholesterol,Glucose_Level,Heart_Rate,Sleep_Hours,Exercise_Hours,Water_Intake,Stress_Level,...,Diet,MentalHealth,PhysicalActivity,MedicalHistory,Allergies,Diet_Type__Vegan,Diet_Type__Vegetarian,Blood_Group_AB,Blood_Group_B,Blood_Group_O
0,2.0,26.0,111.0,198.0,99.0,72.0,4.0,1.0,5.0,5.0,...,1,2,1,0,1,False,True,True,False,False
1,8.0,24.0,121.0,199.0,103.0,75.0,2.0,1.0,2.0,9.0,...,1,2,1,2,2,False,False,True,False,False
2,81.0,27.0,147.0,203.0,100.0,74.0,10.0,-0.0,5.0,1.0,...,2,0,0,1,0,True,False,False,False,False
3,25.0,21.0,150.0,199.0,102.0,70.0,7.0,3.0,3.0,3.0,...,1,2,1,2,0,True,False,False,True,False
4,24.0,26.0,146.0,202.0,99.0,76.0,10.0,2.0,5.0,1.0,...,2,0,2,0,2,False,True,False,True,False


In [55]:
df.isnull().sum()
df.dtypes

Age                      float64
BMI                      float64
Blood_Pressure           float64
Cholesterol              float64
Glucose_Level            float64
Heart_Rate               float64
Sleep_Hours              float64
Exercise_Hours           float64
Water_Intake             float64
Stress_Level             float64
Target                     int64
Smoking                    int64
Alcohol                    int64
Diet                       int64
MentalHealth               int64
PhysicalActivity           int64
MedicalHistory             int64
Allergies                  int64
Diet_Type__Vegan            bool
Diet_Type__Vegetarian       bool
Blood_Group_AB              bool
Blood_Group_B               bool
Blood_Group_O               bool
dtype: object

In [56]:
x=df.drop("Target",axis=1)
y=df["Target"]

In [57]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=42)

In [58]:

from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
x_train_scaled=scaler.fit_transform(x_train)
x_test_scaled=scaler.transform(x_test)

<h1>LOGISTIC REGRESSION</h1>

In [76]:
model1=LogisticRegression(max_iter=1000,penalty='l2',random_state=42,solver="liblinear")
model1.fit(x_train_scaled,y_train)
y_pred=model1.predict(x_test_scaled)
print(f"accuracy{accuracy_score(y_pred,y_test)}")
print(f"recall{recall_score(y_pred,y_test)}")
print(f"clasification_report{classification_report(y_pred,y_test)}")



accuracy0.818848167539267
recall0.8252299605781866
clasification_report              precision    recall  f1-score   support

           0       0.80      0.81      0.81      1343
           1       0.83      0.83      0.83      1522

    accuracy                           0.82      2865
   macro avg       0.82      0.82      0.82      2865
weighted avg       0.82      0.82      0.82      2865



<h1>KNN</h1>

In [84]:
model2 = KNeighborsClassifier(n_neighbors=5,metric="euclidean")
model2.fit(x_train_scaled,y_train)
y_pred=model2.predict(x_test_scaled)
print(f"accuracy {accuracy_score(y_pred,y_test)}")
print(f"recall {recall_score(y_pred,y_test)}")
print(classification_report(y_pred,y_test))


accuracy 0.8659685863874346
recall 0.875250166777852
              precision    recall  f1-score   support

           0       0.86      0.86      0.86      1366
           1       0.87      0.88      0.87      1499

    accuracy                           0.87      2865
   macro avg       0.87      0.87      0.87      2865
weighted avg       0.87      0.87      0.87      2865



<h1>Random forest classifier</h1>

In [97]:
model3=RandomForestClassifier(n_estimators=200)
model3.fit(x_train_scaled,y_train)
y_pred=model3.predict(x_test_scaled)
print(f"accuracy {accuracy_score(y_pred,y_test)}")
print(f"recall {recall_score(y_pred,y_test)}")
print(classification_report(y_pred,y_test))

accuracy 0.937870855148342
recall 0.9375410913872453
              precision    recall  f1-score   support

           0       0.93      0.94      0.93      1344
           1       0.94      0.94      0.94      1521

    accuracy                           0.94      2865
   macro avg       0.94      0.94      0.94      2865
weighted avg       0.94      0.94      0.94      2865



<h1>GradientBoosting<h1>

In [91]:
model4 = GradientBoostingClassifier(n_estimators=1000, learning_rate=1.0,
    max_depth=1, random_state=0).fit(x_train_scaled, y_train)
y_pred=model4.predict(x_test_scaled)
print(f"accuracy {accuracy_score(y_pred,y_test)}")
print(f"recall {recall_score(y_pred,y_test)}")
print(classification_report(y_pred,y_test))

accuracy 0.8610820244328098
recall 0.8695941450432468
              precision    recall  f1-score   support

           0       0.86      0.85      0.85      1362
           1       0.87      0.87      0.87      1503

    accuracy                           0.86      2865
   macro avg       0.86      0.86      0.86      2865
weighted avg       0.86      0.86      0.86      2865



<h1>vootingclassifier<h1>

In [93]:
model5=VotingClassifier(estimators=[
    ("lr",LogisticRegression()),
    ("knn",KNeighborsClassifier()),
    ("randomforest",RandomForestClassifier())
],voting='hard')
model5.fit(x_train_scaled,y_train)
y_pred=model5.predict(x_test_scaled)
print(f"accuracy {accuracy_score(y_pred,y_test)}")
print(f"recall {recall_score(y_pred,y_test)}")
print(classification_report(y_pred,y_test))


accuracy 0.9162303664921466
recall 0.9182597231377719
              precision    recall  f1-score   support

           0       0.91      0.91      0.91      1348
           1       0.92      0.92      0.92      1517

    accuracy                           0.92      2865
   macro avg       0.92      0.92      0.92      2865
weighted avg       0.92      0.92      0.92      2865



## Results

| Model                | Recall |
|----------------------|:------:|
| Logistic Regression  | 82.5%  |
| KNN                  | 87.5%  |
| Random Forest        | 93.7%  |
| Gradient Boosting    | 86.9%  |
| Voting Classifier    | 91.8% |

### Best Classifier that we should use for NovaGen(based on Recall) - Random Forest with accuracy of 93.7%

In [103]:
pip install streamlit pandas numpy scikit-learn matplotlib joblib

   ---------------------------------------- 0.0/9.2 MB ? eta -:--:--
   --- ------------------------------------ 0.8/9.2 MB 3.9 MB/s eta 0:00:03
   ------- -------------------------------- 1.8/9.2 MB 4.6 MB/s eta 0:00:02
   ------------ --------------------------- 2.9/9.2 MB 4.8 MB/s eta 0:00:02
   ----------------- ---------------------- 3.9/9.2 MB 4.8 MB/s eta 0:00:02
   --------------------- ------------------ 5.0/9.2 MB 4.7 MB/s eta 0:00:01
   -------------------------- ------------- 6.0/9.2 MB 4.8 MB/s eta 0:00:01
   ------------------------------ --------- 7.1/9.2 MB 4.7 MB/s eta 0:00:01
   ----------------------------------- ---- 8.1/9.2 MB 4.8 MB/s eta 0:00:01
   ---------------------------------------  9.2/9.2 MB 4.8 MB/s eta 0:00:01
   ---------------------------------------- 9.2/9.2 MB 4.8 MB/s  0:00:01
   ---------------------------------------- 0.0/797.5 kB ? eta -:--:--
   ---------------------------------------- 797.5/797.5 kB 4.4 MB/s  0:00:00
   -----------------------


[notice] A new release of pip is available: 25.3 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [105]:
import joblib

joblib.dump(model3, "health_model.pkl")
joblib.dump(scaler, "scaler.pkl")

['scaler.pkl']